qa/L0_e2e/test_model.py

```diff
@@ -0,0 +1,563 @@
+        assert output_name in shared_memory_regions
+
+    return (triton_input, triton_output, output_handle, input_name, output_name)
+
+
+def get_result(response, output_handle):
+    """Convert Triton response to NumPy array"""
+    if output_handle is None:
+        return response.as_numpy('output__0')
+    else:
+        network_result = TritonMessage(response.get_output('output__0'))
+        return shm.get_contents_as_numpy(
+            output_handle,
+            triton_utils.triton_to_np_dtype(network_result.datatype),
+            network_result.shape
+        )
+
+
+def triton_predict(
+        model_name, arr, model_version='1', protocol='grpc', host='localhost',
+        port=None, shared_mem=None, predict_proba=False, num_classes=1,
+        attempts=3):
+    """Perform prediction on a numpy array using a Triton model"""
+    if port is None:
+        port = STANDARD_PORTS[protocol]
+    client = get_triton_client(protocol=protocol, host=host, port=port)
+    start = perf_counter()
+    try:
+        triton_input, triton_output, handle, input_name, output_name = set_up_triton_io(
+            client,
+            arr,
+            protocol=protocol,
+            shared_mem=shared_mem,
+            predict_proba=predict_proba,
+            num_classes=num_classes
+        )
+
+        result = client.infer(
+            model_name,
+            model_version=str(model_version),
+            inputs=[triton_input],
+            outputs=[triton_output]
+        )
+    except triton_utils.InferenceServerException as exc:
+        # Workaround for Triton Python client bug
+        if exc.status() == 'StatusCode.NOT_FOUND' and attempts > 1:
+            return triton_predict(
+                model_name,
+                arr,
+                model_version=model_version,
+                protocol=protocol,
+                host=host,
+                port=port,
+                shared_mem=shared_mem,
+                predict_proba=predict_proba,
+                num_classes=num_classes,
+                attempts=attempts - 1
+            )
+        raise
+    output = get_result(result, handle)
+    elapsed = perf_counter() - start
+
+    if input_name is not None:
+        client.unregister_cuda_shared_memory(name=input_name)
+    if output_name is not None:
+        client.unregister_cuda_shared_memory(name=output_name)
+
+    return output, elapsed
+
+
+def run_test(
+        model_repo=None,
+        model_name='xgboost_classification_xgboost',
+        model_version=1,
+        protocol='grpc',
+        total_rows=8192,
+        batch_sizes=None,
+        shared_mem=(None, 'cuda'),
+        concurrency=8,
+        timeout=60,
+        retries=3,
+        host='localhost',
+        http_port=STANDARD_PORTS['http'],
+        grpc_port=STANDARD_PORTS['grpc']):
+
+    if batch_sizes is None:
+        batch_sizes = []
+        if total_rows > 128:
+            batch_sizes.append(128)
+        batch_sizes.append(1)
```

Comment:
```suggestion
        batch_sizes = [1]
        if total_rows > 128:
            batch_sizes.append(128)
```

Reply:
This is actually a deliberate choice, though one that deserves a comment. By sandwiching the batch size `1` runs in between the larger runs, it increases the likelihood that the server will have to correctly handle inputs of different sizes in the same (server-side) batch of requests, which seems like an important thing to test. Let me add a comment explaining that.

Reply:
Thanks! I was wondering about why 1 was inserted in the middle of the list.